staging/vhost-device-console/src/vhu_console.rs

```diff
@@ -0,0 +1,1441 @@
+            }
+        }
+
+        Ok(())
+    }
+
+    fn process_ctrl_rx_requests(
+        &mut self,
+        requests: Vec<ConsoleDescriptorChain>,
+        vring: &VringRwLock,
+    ) -> Result<bool> {
+        trace!("process_ctrl_rx_requests");
+        let mut used_flag = false;
+
+        if requests.is_empty() {
+            trace!("Control rx queue is_empty");
+            return Ok(true);
+        }
+
+        for desc_chain in requests {
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+
+            // TODO: Is that test usefull?
+            if descriptors.is_empty() {
+                return Err(Error::UnexpectedDescriptorCount(descriptors.len()));
+            }
+
+            let desc_request = descriptors[0];
+            if !desc_request.is_write_only() {
+                return Err(Error::UnexpectedWriteOnlyDescriptor(0));
+            }
+
+            if (desc_request.len() as usize) < size_of::<VirtioConsoleControl>() {
+                return Err(Error::UnexpectedDescriptorSize(
+                    size_of::<VirtioConsoleControl>(),
+                    desc_request.len(),
+                ));
+            }
+
+            let ctrl_msg: VirtioConsoleControl = match self.rx_ctrl_fifo.remove() {
+                Ok(item) => {
+                    used_flag = true;
+                    item
+                }
+                _ => {
+                    trace!("No rx elements");
+                    return Ok(used_flag);
+                }
+            };
+
+            self.print_console_frame(ctrl_msg);
+
+            let mut buffer: Vec<u8> = Vec::new();
+            buffer.extend_from_slice(&ctrl_msg.to_le_bytes());
+
+            if ctrl_msg.event.to_native() == VIRTIO_CONSOLE_PORT_NAME {
+                buffer.extend_from_slice(b"org.test.foo!");
+            };
+
+            desc_chain
+                .memory()
+                .write_slice(&buffer, desc_request.addr())
+                .map_err(|_| Error::DescriptorWriteFailed)?;
+
+            if vring
+                .add_used(desc_chain.head_index(), desc_request.len())
+                .is_err()
+            {
+                warn!("Couldn't return used descriptors to the ring");
+            }
+        }
+
+        Ok(true)
+    }
+
+    fn handle_control_msg(&mut self, ctrl_msg: VirtioConsoleControl) -> Result<()> {
+        let mut ctrl_msg_reply = VirtioConsoleControl {
+            id: 0.into(),
+            event: 0.into(),
+            value: 1.into(),
+        };
+        match ctrl_msg.event.to_native() {
+            VIRTIO_CONSOLE_DEVICE_READY => {
+                trace!("VIRTIO_CONSOLE_DEVICE_READY");
+                self.ready = true;
+                ctrl_msg_reply.event = VIRTIO_CONSOLE_PORT_ADD.into();
+                self.rx_ctrl_fifo
+                    .add(ctrl_msg_reply)
+                    .expect("Control message could not be added into queue");
```

Comment:
is it intended to panic here? can the device not recover from this error? 
if that's not the case, I will suggest to use the `map_err` method to properly propagate the error.